src/tla/__rewire_finite_sets_ext_in_apalache.tla

```diff
@@ -0,0 +1,166 @@
+ *
+ * @type: ((a, b) => b, b, Set(a)) => b;
+ *)
+FoldSet(op(_, _), base, set) ==
+    \* ApalacheFoldSet is accumulating the result in the left argument,
+    \* whereas FinitSetsExt!FoldSet is accumulating the result
+    \* in the right argument.
+    LET OpSwap(x, y) == op(y, x) IN
+    __ApalacheFoldSet(OpSwap, base, set)
+
+(**
+ * Calculate the sum of the elements in set.
+ *
+ * Example: Sum(0 .. 10) = 55
+ *
+ * @type: Set(Int) => Int;
+ *)
+SumSet(set) ==
+   LET Plus(x, y) == x + y IN
+   __ApalacheFoldSet(Plus, 0, set)
+
+(**
+ * Calculuate the product of the elements in set.
+ *
+ * Example: Product(1 .. 3) = 6
+ *
+ * @type: Set(Int) => Int;
+ *)
+ProductSet(set) ==
+   LET Mult(x, y) == x * y IN
+   __ApalacheFoldSet(Mult, 1, set)
+
+(**
+ * An alias for FoldSet. ReduceSet was used instead of FoldSet in
+ * earlier versions of the community modules.
+ *
+ * @type: ((a, b) => b, Set(a), b) => b;
+ *)
+ReduceSet(op(_, _), set, acc) == 
+   FoldSet(op, set, acc)
+
+(**
+ * Starting from base, apply op to f(x), for all x \in S, in an arbitrary
+ * order. It is assumed that op is associative and commutative.
+ *
+ * Example: FlattenSet({{1},{2}}) = {1,2}
+ *
+ * @type: Set(Set(a)) => Set(a);
+ *)
+FlattenSet(S) ==
+  LET SetUnion(A, B) == A \union B IN
+  __ApalacheFoldSet(SetUnion, {}, S)
+
+(**
+ * The symmetric difference of two sets.
+ *
+ * The symmetric difference of sets A and B is the set containing all
+ * elements that are present in either A or B but not in their intersection.
+ *
+ * @type: (Set(a), Set(a)) => Set(a);
+ *)
+SymDiff(A, B) ==
+    { x \in A: x \notin B } \union { x \in B: x \notin A }
+
+(**
+ * Quantify the elements in S matching the predicate (LAMDBA) P.
+ *
+ * This operator is overridden by FiniteSetsExt#quantify whose
+ * implementation does *not* enumerate the intermediate set! This is
+ * the only advantage that Quantify(...) has over Cardinality(...).
+ *
+ * Example: Quantify(1..9, LAMBDA n : n % 2 = 0) = 4
+ *
+ * @type: (Set(a), (a => Bool)) => Int;
+ *)
+Quantify(S, P(_)) ==
+   Cardinality({s \in S : P(s)})
+
+
+(**
+ * A k-subset ks of a set S has Cardinality(ks) = k.  The number of
+ * k-subsets of a set S with Cardinality(S) = n is given by the binomial
+ * coefficients n over k.  A set S with Cardinality(S) = n has 2^n
+ * k-subsets.  \A k \notin 0..Cardinality(S): kSubset(k, S) = {}.
+ *
+ * This operator is overridden by FiniteSetsExt#kSubset whose
+ * implementation, contrary to  { s \in SUBSET S : Cardinality(s) = k },
+ * only enumerates the k-subsets of S and not all subsets.
+ *
+ * Example: kSubset(2, 1..3) = {{1,2},{2,3},{3,1}}
+ *
+ * @type: (Int, Set(a)) => Set(Set(a));
+ *)
+kSubset(k, S) == 
+   { s \in SUBSET S : Cardinality(s) = k }
+
+(**
+ * We define Max(S) and Min(S) to be the maximum and minimum,
+ * respectively, of a finite, non-empty set S of integers.
+ *
+ * @type: Set(Int) => Int;
+ *)
+Max(S) == CHOOSE x \in S : \A y \in S : x >= y
```

Comment:
Does this definition require `S` to be finite?

Reply:
In pure TLA+, `S` can be infinite. Obviously, TLC and Apalache would reject it.